File: 2024/01/solution.py
from collections import Counter

with open('./input', 'r') as f:
    data = f.readlines()

left_list = []
right_list = []
distances = 0

for line in data:
    left, right = line.strip().split("   ")
    left_list.append(int(left))
    right_list.append(int(right))

left_list.sort()
right_list.sort()

# Part 1:
for i in range(len(left_list)):
    distances += abs(left_list[i] - right_list[i])

print(distances)

# Part 2:
p2_score = 0
freqs_right = Counter(right_list)
p2_score = sum(num * freqs_right[num] for num in left_list)
print(p2_score)

File: 2024/01/solution.ex
defmodule AOC2024.Day1 do
  # Parse input directly from a file or pre-split lines
  def parse_input(file_path) do
    File.read!(file_path)
    |> String.split("\n", trim: true)
    |> parse_lines()
  end

  def parse_lines(lines) do
    lines
    |> Enum.map(fn line ->
      Regex.split(~r/\s+/, line)
      |> Enum.map(&String.to_integer/1)
      |> List.to_tuple()
    end)
    |> Enum.unzip()
  end
end

defmodule AOC2024.Day1.Part1 do
  def sum_distance({left, right}) do
    left_sorted = Enum.sort(left)
    right_sorted = Enum.sort(right)

    Enum.zip(left_sorted, right_sorted)
    |> Enum.map(fn {a, b} -> abs(a - b) end)
    |> Enum.sum()
  end
end

defmodule AOC2024.Day1.Part2 do
  def similarity_score({left, right}) do
    frequencies = Enum.frequencies(right)

    Enum.reduce(left, 0, fn el, acc ->
      acc + el * Map.get(frequencies, el, 0)
    end)
  end
end

IO.puts(AOC2024.Day1.Part1.sum_distance(AOC2024.Day1.parse_input("./input")))
IO.puts(AOC2024.Day1.Part2.similarity_score(AOC2024.Day1.parse_input("./input")))